### PR TITLE
add an example to move an elmo

### PR DESCRIPTION
**Pull request by Eyusd**
This MR adds the basic files and implements the CANOpen DSP402 protocol to interact with an Elmo.
It can be started from examples/moveElmo, and simply applies a small torque to it.

All the concepts are here, from the CANOpen FSM to the structure of input and output PDO.

As the specs were loosely defined, this will certainly change when requirements evolve.

### CODE REVIEW

File: examples/elmoControl/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+if (PIKEOS)
+    set(OS_SRCS ${PROJECT_SOURCE_DIR}/src/OS/PikeOS/p4ext_config.c)
+endif()
+
+add_executable(elmoControl elmoControl.cc src/CanOpenStateMachine.cc ${OS_SRCS})
+target_include_directories(elmoControl PUBLIC  ${CMAKE_CURRENT_SOURCE_DIR}/include)
+target_link_libraries(elmoControl kickcat)
+set_kickcat_properties(elmoControl)
```
Comment (Siviuze) [inline examples/elmoControl/CMakeLists.txt:8]: add newline at the end

File: examples/elmoControl/elmoControl.cc [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    Bus bus(link);
+
+
+    uint8_t io_buffer[2048];
+    try
+    {
+        bus.init();
+
+        // Prepare mapping for elmo
+        uint32_t Rx_length = sizeof(hal::pdo::elmo::RxMapping);
```
Comment (Siviuze) [inline examples/elmoControl/elmoControl.cc:79]: rx_length (lower case)

File: examples/elmoControl/elmoControl.cc [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        bus.writeSDO(bus.slaves().at(0), hal::sdo::elmo::RxPDO.index, hal::sdo::elmo::RxPDO.subindex, 
+                    Bus::Access::COMPLETE, static_cast<void*>(const_cast<uint8_t*>(hal::pdo::elmo::RxMapping)), Rx_length);
+        uint32_t Tx_length = sizeof(hal::pdo::elmo::TxMapping);
```
Comment (Siviuze) [inline examples/elmoControl/elmoControl.cc:82]: lower case

File: examples/elmoControl/elmoControl.cc [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    uint8_t io_buffer[2048];
+    try
+    {
+        bus.init();
+
+        // Prepare mapping for elmo
+        uint32_t Rx_length = sizeof(hal::pdo::elmo::RxMapping);
+        bus.writeSDO(bus.slaves().at(0), hal::sdo::elmo::RxPDO.index, hal::sdo::elmo::RxPDO.subindex, 
+                    Bus::Access::COMPLETE, static_cast<void*>(const_cast<uint8_t*>(hal::pdo::elmo::RxMapping)), Rx_length);
```
Comment (Siviuze) [inline examples/elmoControl/elmoControl.cc:81]: static_cast is useless

File: examples/elmoControl/include/CanOpenErrors.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
+#ifndef CAN_ELMO_ERROR_H
+#define CAN_ELMO_ERROR_H
+
+#include <cstdint>
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenErrors.h:4]: not sure why theses includes are requiered

File: examples/elmoControl/include/CanOpenErrors.h [hunk 6]
```diff
@@ -0,0 +1,88 @@
+#ifndef CAN_ELMO_ERROR_H
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenErrors.h:1]: Prefix with KICKCAT_

File: examples/elmoControl/include/CanOpenErrors.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+#define CAN_ELMO_ERROR_H
+
+#include <cstdint>
+#include <cstdio>
+
+namespace can
+{
+    namespace emergency
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenErrors.h:9]: too many namespaces and not meaningful. BTW the first namespace shall be kickcat

File: examples/elmoControl/include/CanOpenStateMachine.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
+#ifndef CAN_ELMO_STATE_MACHINE_H
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenStateMachine.h:1]: prefix with KICKCAT_

File: examples/elmoControl/include/CanOpenStateMachine.h [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+#define CAN_ELMO_STATE_MACHINE_H
+
+#include <cstdint>
+#include <chrono>
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenStateMachine.h:5]: include kickcat Time.h

File: examples/elmoControl/include/CanOpenStateMachine.h [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+
+namespace can
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenStateMachine.h:8]: first namespace shall be kickcat

File: examples/elmoControl/include/CanOpenStateMachine.h [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+{
+    /// \enum CANOpenState The different states of a CANOpen motor.
+    enum class CANOpenState
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenStateMachine.h:11]: This seems to be a DS402 specific state (and not just CANOpen)

File: examples/elmoControl/include/CanOpenStateMachine.h [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            NO_MODE    = -1, // Default value to mean "error occurred"
+            POSITION   = 1,  // Profiled position (point to point) mode
+            VELOCITY   = 3,  // Profiled velocity mode
+            TORQUE     = 4,   // Profiled torque mode
+            POSITION_CYCLIC = 8 // Direct position control without ramps
+        };
+
+    }
+
+    using namespace std::chrono;
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenStateMachine.h:94]: Time.h

File: examples/elmoControl/include/ElmoProtocol.h [hunk 13]
```diff
@@ -0,0 +1,149 @@
+#ifndef CAN_ELMO_PROTOCOL_H
+#define CAN_ELMO_PROTOCOL_H
+
+#include <cstdint>
+
+struct SDOEntry
```
Comment (Siviuze) [inline examples/elmoControl/include/ElmoProtocol.h:6]: Note sure why this is needed

File: examples/elmoControl/include/ElmoProtocol.h [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+{
+    uint16_t index;         ///< Main index in the SDO dict
+    uint8_t  subindex;      ///< Subindex in the SDO dict
+    char const* name{""};   ///< Name of SDO, only for print purpose
+};
+
+namespace hal
```
Comment (Siviuze) [inline examples/elmoControl/include/ElmoProtocol.h:13]: kickcat again, hal has no meaning here

File: examples/elmoControl/include/ElmoProtocol.h [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+{
+    typedef float           float32_t;
```
Comment (Siviuze) [inline examples/elmoControl/include/ElmoProtocol.h:15]: avoid this (use float directly)

File: examples/elmoControl/include/ElmoProtocol.h [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+                uint32_t velocityOffset;
+                int32_t  digitalOutput;
+            } __attribute__((packed));
+            constexpr uint8_t RxMapping[] = { 0x07, 0x00, 0x0A, 0x16, 0x0B, 0x16, 0x0C, 0x16, 0x0D, 0x16, 0x0F, 0x16, 0x17, 0x16, 0x1D, 0x16 };
+            //TODO TorqueOffset: add torque offset field: int16_t {0x18, 0x16} and increment RxMapping count from 0x07 to 0x08
+        }
+
+    }
+
+    namespace sdo
```
Comment (Siviuze) [inline examples/elmoControl/include/ElmoProtocol.h:56]: seems a lot are unused?

File: examples/elmoControl/include/ElmoProtocol.h [hunk 17]
```diff
@@ -0,0 +1,149 @@
+#ifndef CAN_ELMO_PROTOCOL_H
```
Comment (Siviuze) [inline examples/elmoControl/include/ElmoProtocol.h:1]: KICKCAT_

File: examples/elmoControl/src/CanOpenStateMachine.cc [hunk 18]
```diff
@@ -0,0 +1,92 @@
+#include "CanOpenStateMachine.h"
+#include "kickcat/Time.h"
+#include "kickcat/Error.h"
+
+namespace can
```
Comment (Siviuze) [inline examples/elmoControl/src/CanOpenStateMachine.cc:5]: kickcat

**Comment by Siviuze** [comment]
Seems like some files should be added to the library directly and not just in the example (like state machine and CAN errors codes)

File: examples/elmoControl/elmoControl.cc [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+#ifdef __linux__
+    #include "kickcat/OS/Linux/Socket.h"
+#elif __PikeOS__
+    #include "kickcat/OS/PikeOS/Socket.h"
+#else
+    #error "Unknown platform"
+#endif
+
+#include <iostream>
+#include <fstream>
```
Comment (Siviuze) [inline examples/elmoControl/elmoControl.cc:19]: not sure why fsteam and algorithm are used for.

File: examples/elmoControl/include/ElmoProtocol.h [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+{
+    struct SDOEntry
+    {
+        uint16_t index;         ///< Main index in the SDO dict
+        uint8_t  subindex;      ///< Subindex in the SDO dict
+    };
+
+    namespace sdo
+    {
+        constexpr SDOEntry RxPDO { 0x1C12, 0 };
```
Comment (Siviuze) [inline examples/elmoControl/include/ElmoProtocol.h:14]: Init here with values from protocol.h `CoE::SM_CHANNEL + 2` (fell free to add SM_CHANNEL_0 etc)

File: examples/elmoControl/src/CanOpenStateMachine.cc [hunk 21]
```diff
@@ -0,0 +1,92 @@
+#include "CanOpenStateMachine.h"
+#include "kickcat/Time.h"
+#include "kickcat/Error.h"
+
+namespace kickcat
+{
+    void CANOpenStateMachine::update(uint16_t statusWord)
+        {
```
Comment (Siviuze) [inline examples/elmoControl/src/CanOpenStateMachine.cc:8]: indentation

File: examples/elmoControl/src/CanOpenStateMachine.cc [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+                        default:
+                        case CANOpenState::ON: DEBUG_PRINT("ON status achieved\n"); break;
+                        case CANOpenState::FAULT:
+                            // Do nothing
+                            break;
+                    }
+
+                    // Time out, motor start failed
+                    if (motorState_ != CANOpenState::ON
+                        && motorState_ != CANOpenState::FAULT
```
Comment (Siviuze) [inline examples/elmoControl/src/CanOpenStateMachine.cc:63]: replace && with and

File: examples/elmoControl/include/ElmoProtocol.h (hunk 20 above)
Comment (Siviuze) [reply examples/elmoControl/include/ElmoProtocol.h:14]: BTW not sure if this abstraction if of any use: may be more useful to have the constant directly in the main.cc 

```
bus.writeSDO(bus.slaves().at(0), 0x1C12, 0, Bus::Access::COMPLETE, 
     const_cast<uint8_t*>(pdo::RxMapping), rxLength);
```

File: examples/elmoControl/include/CanOpenStateMachine.h [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+    // Timeouts to prevent blocking the statemachine in case of failure
+    constexpr nanoseconds MOTOR_RESET_DELAY = 10ms;
+    constexpr nanoseconds MOTOR_INIT_TIMEOUT = 1s;
+
+    class CANOpenStateMachine
+    {
+    public:
+        CANOpenCommand command_  = CANOpenCommand::NONE;
+        CANOpenState motorState_ = CANOpenState::OFF;
```
Comment (Siviuze) [inline examples/elmoControl/include/CanOpenStateMachine.h:100]: snake case

File: examples/elmoControl/include/ElmoProtocol.h [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+        constexpr SDOEntry TxPDO { 0x1C13, 0 };
+    }
+
+    namespace pdo
+    {
+        /// \details See CANOpen DS402 document for description of fields.
+        ///          Here fields are name "Torque" but actually refer to current in amps.
+        struct Input
+        {
+            uint16_t statusWord;
```
Comment (Siviuze) [inline examples/elmoControl/include/ElmoProtocol.h:24]: snake_case everywhere

File: examples/elmoControl/elmoControl.cc [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+    catch (std::exception const& e)
+    {
+        std::cerr << e.what() << std::endl;
+        return 1;
+    }
+
+    link->setTimeout(500us);
+
+    Slave& elmo = bus.slaves().at(0);
+    pdo::Output* outputPDO = reinterpret_cast<pdo::Output*>(elmo.output.data);
```
Comment (Siviuze) [inline examples/elmoControl/elmoControl.cc:130]: snake case (maybe remove PDO)

Comment (Eyusd) [reply examples/elmoControl/elmoControl.cc:130]: namespace pdo prevents mixup with enum kickcat::SyncManagerType (containing Output and Input)

File: examples/elmo_control/include/ElmoProtocol.h [hunk 26]
```diff
@@ -0,0 +1,47 @@
+#ifndef KICKCAT_CAN_ELMO_PROTOCOL_H
+#define KICKCAT_CAN_ELMO_PROTOCOL_H
+
+namespace kickcat
+{
+    struct SDOEntry
```
Comment (Siviuze) [inline examples/elmo_control/include/ElmoProtocol.h:6]: used nowhere

File: examples/elmo_control/include/CanOpenStateMachine.h [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+    constexpr nanoseconds MOTOR_RESET_DELAY = 10ms;
+    constexpr nanoseconds MOTOR_INIT_TIMEOUT = 1s;
+
+    class CANOpenStateMachine
+    {
+    public:
+        CANOpenCommand command_  = CANOpenCommand::NONE;
+        CANOpenState motor_state_ = CANOpenState::OFF;
+        nanoseconds start_motor_timestamp_ = 0ns;
+        uint16_t control_word_ = 0;
```
Comment (Siviuze) [inline examples/elmo_control/include/CanOpenStateMachine.h:102]: public member?

**Review by Siviuze — APPROVED** [review]

**Pull request merged**